### PR TITLE
bug: update terraform syntax for working directory

### PR DESCRIPTION
**Pull request by JayCeeJr**
Related issue: https://github.com/go-vela/community/issues/389

### CODE REVIEW

File: cmd/vela-terraform/apply.go [hunk 1]
```diff
@@ -138,7 +143,7 @@ func (a *Apply) Command(dir string) *exec.Cmd {
 	}
 
 	// add the required dir param
-	flags = append(flags, dir)
+	flags = append(flags)
```
Comment (JordanSussman) [inline cmd/vela-terraform/apply.go:146]: This should be removed since nothing is being appended.

File: cmd/vela-terraform/destroy.go [hunk 2]
```diff
@@ -130,7 +135,7 @@ func (a *Destroy) Command(dir string) *exec.Cmd {
 	}
 
 	// add the required dir param
-	flags = append(flags, dir)
+	flags = append(flags)
```
Comment (JordanSussman) [inline cmd/vela-terraform/destroy.go:138]: This should be removed since nothing is being appended.

File: cmd/vela-terraform/fmt.go [hunk 3]
```diff
@@ -61,7 +66,7 @@ func (f *FMT) Command(dir string) *exec.Cmd {
 	}
 
 	// add the required dir param
-	flags = append(flags, dir)
+	flags = append(flags)
```
Comment (JordanSussman) [inline cmd/vela-terraform/fmt.go:69]: This should be removed since nothing is being appended.

File: cmd/vela-terraform/init.go [hunk 4]
```diff
@@ -162,7 +167,7 @@ func (i *Init) Command(dir string) *exec.Cmd {
 	}
 
 	// add the required dir param
-	flags = append(flags, dir)
+	flags = append(flags)
```
Comment (JordanSussman) [inline cmd/vela-terraform/init.go:170]: This should be removed since nothing is being appended.

File: cmd/vela-terraform/plan.go [hunk 5]
```diff
@@ -146,7 +151,7 @@ func (p *Plan) Command(dir string) *exec.Cmd {
 	}
 
 	// add the required dir param
-	flags = append(flags, dir)
+	flags = append(flags)
```
Comment (JordanSussman) [inline cmd/vela-terraform/plan.go:154]: This should be removed since nothing is being appended.

File: cmd/vela-terraform/validate.go [hunk 6]
```diff
@@ -65,7 +70,7 @@ func (v *Validation) Command(dir string) *exec.Cmd {
 	}
 
 	// add the required dir param
-	flags = append(flags, dir)
+	flags = append(flags)
```
Comment (JordanSussman) [inline cmd/vela-terraform/validate.go:73]: This should be removed since nothing is being appended.

**Review by JordanSussman — CHANGES_REQUESTED** [review]
remove append when nothing is appended

**Review by wass3r — COMMENTED** [review]
Thanks for the PR.

Is this backwards compatible for when users change or use versions to pre v1.0? (https://github.com/go-vela/vela-terraform/blob/master/cmd/vela-terraform/main.go#L295-L305)?

EDIT: looks like it was added in 0.14+ (https://github.com/hashicorp/terraform/commit/efe78b2910c5b1f2292f0a16d990b2ad92352feb)? so the plugin would break for users of versions prior to that?

**Comment by JayCeeJr** [comment]
@wass3r I added backwards compatibility

File: cmd/vela-terraform/main.go [hunk 7]
```diff
@@ -292,13 +294,19 @@ func run(c *cli.Context) error {
 		"registry": "https://hub.docker.com/r/target/vela-terraform",
 	}).Info("Vela Terraform Plugin")
 
-	// capture custom terraform version requested
-	version := c.String("terraform.version")
+	// capture custom terraform tfVersion requested
+	tfVersion := c.String("terraform.tfVersion")
```
Comment (wass3r) [inline cmd/vela-terraform/main.go:298]: could be misunderstanding but why did we change this to `terraform.tfVersion`. it looks like it's still named `terraform.version` here: https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/cmd/vela-terraform/main.go#L147-L152

**Review by wass3r — CHANGES_REQUESTED** [review]
thanks for adding that backwards compatibility 🎉 

i added one question for the changes.

File: cmd/vela-terraform/main.go (hunk 7 above)
Comment (JayCeeJr) [reply cmd/vela-terraform/main.go:298]: I refactored that variable because of a name collision with the import "github.com/go-vela/vela-terraform/version"

Comment (wass3r) [reply cmd/vela-terraform/main.go:298]: i think the golang variable rename `version` -> `tfVersion` is not of concern. but the Flag name reference of `terraform.tfVersion` is not accurate anymore, is it?

Comment (JayCeeJr) [reply cmd/vela-terraform/main.go]: Nice catch, I fixed that.

**Review by wass3r — APPROVED** [review]

File: cmd/vela-terraform/main.go [hunk 8]
```diff
@@ -292,13 +294,19 @@ func run(c *cli.Context) error {
 		"registry": "https://hub.docker.com/r/target/vela-terraform",
 	}).Info("Vela Terraform Plugin")
 
-	// capture custom terraform version requested
-	version := c.String("terraform.version")
+	// capture custom terraform tfVersion requested
+	tfVersion := c.String("terraform.version")
+
+	tfSemVersion, err := semver.NewVersion(tfVersion)
```
Comment (JordanSussman) [inline cmd/vela-terraform/main.go:300]: Need to handle the scenario where tfVersion is `""` since the user opted to utilize the pre installed terrafrom binary.

**Review by JordanSussman — CHANGES_REQUESTED** [review]
Handle empty tfVersion variable.

File: cmd/vela-terraform/main.go (hunk 8 above)
Comment (wass3r) [reply cmd/vela-terraform/main.go:300]: we could shove https://github.com/go-vela/vela-terraform/blob/a00b059b732a3768b86de12364912b44bf5e253f/cmd/vela-terraform/main.go#L300-L304 into https://github.com/go-vela/vela-terraform/blob/a00b059b732a3768b86de12364912b44bf5e253f/cmd/vela-terraform/main.go#L307 . but, i think we're fine probably?

File: cmd/vela-terraform/main.go [hunk 9]
```diff
@@ -292,16 +294,24 @@ func run(c *cli.Context) error {
...
+	// check if a custom terraform tfVersion was requested
+	if len(tfVersion) > 0 {
+		// attempt to install the custom terraform tfVersion
+		err := install(tfVersion, os.Getenv("PLUGIN_TERRAFORM_VERSION"))
 		if err != nil {
 			return err
 		}
+	} else {
+		// Default Version will be used if not defined
+		tfVersion = "1.0.0"
```
Comment (JayCeeJr) [inline cmd/vela-terraform/main.go:309]: If no version is specified, this will avoid the additional download and use the 1.0 behavior by default.

Comment (wass3r) [reply cmd/vela-terraform/main.go:309]: Do we need this? We're already setting a "default" here: https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/Dockerfile#L8

That's the "default" version bundled with the image. If no custom version was defined, it uses that one.

---

FWIW, that value is also set here: https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/Dockerfile#L38

And then used here: https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/cmd/vela-terraform/main.go#L301 to avoid a download if someone specifies the same version the image was built with.

Comment (JordanSussman) [reply cmd/vela-terraform/main.go:309]: > Do we need this? We're already setting a "default" here:

This is incorrect since the value of `c.String("terraform.version")` is only being populated via the environment variables of [PARAMETER_VERSION and TERRAFORM_VERSION](https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L148).

Comment (wass3r) [reply cmd/vela-terraform/main.go:309]: hmm, maybe i'm missing something.

if `terraform.version` wasn't given it will skip the install and use the bundled "default" version that lives at https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/cmd/vela-terraform/terraform.go#L20 which is made available when the image is built.

if we're just talking about deriving the version value, can we just reference `PLUGIN_TERRAFORM_VERSION`'s value instead? otherwise, we are hard-coding default version values in a number of places now.

Comment (JordanSussman) [reply cmd/vela-terraform/main.go:309]: > if we're just talking about deriving the version value, can we just reference PLUGIN_TERRAFORM_VERSION's value instead? otherwise, we are hard-coding default version values in a number of places now.

The current code changes requires that `terraform.version` is populated to allow the [SupportsChdir](https://github.com/go-vela/vela-terraform/blob/0bb13edc75df3c5664437e993839704773f8abd8/cmd/vela-terraform/main.go#L415) function to determine what logic to utilize. Do you think we should add `PLUGIN_TERRAFORM_VERSION` to the front of [this line](https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L148) to allow the user to override the default?

Comment (JayCeeJr) [reply cmd/vela-terraform/main.go:309]: IIRC: The ARG line in the Dockerfile is only available during the build.

Comment (wass3r) [reply cmd/vela-terraform/main.go:309]: the ARG is, but we are populating the ENV with it in a later stage in the Dockerfile
```
✦ ❯ docker run -it --entrypoint sh target/vela-terraform
/ # env
HOSTNAME=42365df4871e
SHLVL=1
HOME=/root
PLUGIN_TERRAFORM_VERSION=1.0.0
TERM=xterm
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
/ # 
```

@JordanSussman sorry, for delay. if `terraform.version` isn't supplied, we could use `PLUGIN_TERRAFORM_VERSION` - wouldn't that work?

Comment (JordanSussman) [reply cmd/vela-terraform/main.go:309]: > @JordanSussman sorry, for delay. if terraform.version isn't supplied, we could use PLUGIN_TERRAFORM_VERSION - wouldn't that work?

Are you saying that we should add `PLUGIN_TERRAFORM_VERSION` to the list of parsed environment variables that urfavecli utilizes to set the value of `terraform.version`?

https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L147-L152

I would be happy to append `PLUGIN_TERRAFORM_VERSION` to the end of the existing list.

Comment (wass3r) [reply cmd/vela-terraform/main.go:309]: As long as the user supplied value overrides it, i don't see a downside?

Comment (JayCeeJr) [reply cmd/vela-terraform/main.go]: Check https://github.com/JayCeeJr/vela-terraform/commit/d72ed96ed8617e06f68d6001b4c6ff83e6985cce
I think this will work and use the default version if not specified

Comment (JordanSussman) [reply cmd/vela-terraform/main.go]: @JayCeeJr you should be able to remove https://github.com/JayCeeJr/vela-terraform/commit/d72ed96ed8617e06f68d6001b4c6ff83e6985cce in favor of adding `PLUGIN_TERRAFORM_VERSION` to the beginning of https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L148.

Comment (JayCeeJr) [reply cmd/vela-terraform/main.go]: @JordanSussman , I'm fairly certain we don't want to do that.

If we do that won't `c.String("terraform.version")  = os.Getenv("PLUGIN_TERRAFORM_VERSION")`
https://github.com/go-vela/vela-terraform/blob/d72ed96ed8617e06f68d6001b4c6ff83e6985cce/cmd/vela-terraform/main.go#L298-L304

Which will make this always true 
https://github.com/go-vela/vela-terraform/blob/8930ba57f593c8731699d231198bf12fe62658e1/cmd/vela-terraform/terraform.go#L32-L36
and never install a new version.

That being said, maybe just setting the default for the arg to match the preinstalled version will skip the install anyway.

Comment (JayCeeJr) [reply cmd/vela-terraform/main.go]: Are the EnvVars applied in order if multiples are specified?... I guess that would be the same as what I was suggesting.

Comment (JordanSussman) [reply cmd/vela-terraform/main.go]: > Are the EnvVars applied in order if multiples are specified?... I guess that would be the same as what I was suggesting.

I wrote a brief script to validate the behavior of urfavecli:

```golang
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/urfave/cli/v2"
)

func main() {
	app := &cli.App{
		Name:  "boom",
		Usage: "make an explosive entrance",
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:    "test",
				EnvVars: []string{"TEST_VAR_1", "TEST_VAR_2", "TEST_VAR_3"},
			},
		},
		Action: func(c *cli.Context) error {
			fmt.Println(c.String("test"))
			return nil
		},
	}

	err := app.Run(os.Args)
	if err != nil {
		log.Fatal(err)
	}
}
```

```sh
TEST_VAR_2=test2 TEST_VAR_3=test3 TEST_VAR_1=test1 go run main.go
```

TL;DR urfave will utilize the first environment variable within the EnvVars slice that is populated, so you should be safe to append `PLUGIN_TERRAFORM_VERSION` to the end of the existing slice.

Comment (JayCeeJr) [reply cmd/vela-terraform/main.go]: Added to EnvVars
https://github.com/go-vela/vela-terraform/blob/6d75345867705b86b776604e6fc4a22e523efd96/cmd/vela-terraform/main.go#L149-L154

**Review by wass3r — APPROVED** [review]
Thanks for all the work again! ❤️

**Comment by wass3r** [comment]
@JordanSussman it still has you marked as requesting changes.. can you mark as approved if you're good with this, please ;D

**Review by JordanSussman — APPROVED** [review]

**Review by jbrockopp — APPROVED** [review]
LGTM
